consensus/exit_probability_randomizer/walker.c

```diff
@@ -0,0 +1,320 @@
+        // set
+      }
+    }
+  }
+  return RC_OK;
+}
+
+retcode_t select_approver(const ep_randomizer_t *exit_probability_randomizer,
+                          cw_map_t cw_ratings, hash_set_t *approvers,
+                          trit_array_t *approver, bool *has_next_approver) {
+  retcode_t ret = RC_OK;
+  hash_entry_t *curr_approver = NULL;
+  hash_entry_t *tmp_approver = NULL;
+  cw_entry_t *curr_rating = NULL;
+  size_t num_approvers = HASH_COUNT(*approvers);
+  int64_t weights[num_approvers];
+  double sum_weights = 0;
+  size_t idx = 0;
+
+  *has_next_approver = false;
+
+  HASH_ITER(hh, *approvers, curr_approver, tmp_approver) {
+    HASH_FIND(hh, cw_ratings, curr_approver->hash, FLEX_TRIT_SIZE_243,
+              curr_rating);
+    if (!curr_rating) {
+      // TODO - should never happen, figure error
+    }
+    weights[idx++] = curr_rating->cw;
+  }
+
+  int64_t max_weight = 0;
+  for (idx = 0; idx < num_approvers; ++idx) {
+    if (weights[idx] > max_weight) {
+      max_weight = weights[idx];
+    }
+  }
+
+  for (idx = 0; idx < num_approvers; ++idx) {
+    weights[idx] -= max_weight;
+  }
+
+  for (idx = 0; idx < num_approvers; ++idx) {
+    weights[idx] = exp(weights[idx] * exit_probability_randomizer->alpha);
+    sum_weights += weights[idx];
+  }
+
+  double target = ((double)rand() / (double)RAND_MAX) * sum_weights;
+  idx = 0;
+  while (target > 0 && idx < num_approvers) {
+    target -= weights[idx++];
+  }
+  --idx;
+
+  size_t iter_idx = 0;
+  HASH_ITER(hh, *approvers, curr_approver, tmp_approver) {
+    if (iter_idx++ == idx) {
+      *has_next_approver = true;
+      memcpy(approver->trits, curr_approver->hash, FLEX_TRIT_SIZE_243);
+      approver->num_bytes = FLEX_TRIT_SIZE_243;
+      break;
+    }
+  }
+
+  return RC_OK;
+}
+
+retcode_t find_tail_if_valid(const ep_randomizer_t *exit_probability_randomizer,
+                             const trit_array_t *tx_hash,
+                             trit_array_t *tail_hash, bool *found_tail) {
+  retcode_t res;
+
+  iota_stor_pack_t tx_pack;
+
+  tx_pack.capacity = 1;
+  tx_pack.num_loaded = 0;
+  tx_pack.insufficient_capacity = false;
+  struct _iota_transaction curr_tx_s;
+  struct _iota_transaction next_tx_s;
+  iota_transaction_t curr_tx = &curr_tx_s;
+  iota_transaction_t next_tx = &next_tx_s;
+  tx_pack.models = (void **)(&curr_tx);
+  trit_array_t curr_tx_hash;
+  flex_trit_t bundle_hash[FLEX_TRIT_SIZE_243];
+  bool found_approver = false;
+  *found_tail = false;
+
+  res = iota_tangle_transaction_load(exit_probability_randomizer->tangle,
+                                     TRANSACTION_COL_HASH, tx_hash, &tx_pack);
+
+  if (res != RC_OK || tx_pack.num_loaded == 0) {
+    return res;
+  }
+
+  curr_tx = (iota_transaction_t)(tx_pack.models[0]);
+
+  uint32_t index = curr_tx->current_index;
+  memcpy(bundle_hash, curr_tx->bundle, FLEX_TRIT_SIZE_243);
+
+  iota_stor_pack_t hash_pack;
+  if ((res = hash_pack_init(&hash_pack, 10)) != RC_OK) {
```

Comment:
10 ?

Reply:
good point :)
Actually, this value should be the expected value of direct approvers + some lieway, or something like that, so we can avoid resizes,
but for now i will go with some hard coded value